Add support for whitelist (resolves #10, #53)

jessebrennan opened the pull request:
This adds whitelist support for the dashboard.

**Note**: This depends on a https://github.com/DataBiosphere/cgp-deployment/pull/59

There are several sets of changes in this PR that got it to this stage.

1. First the database dependency was removed. Instead most of the important information that was previously there is now stored in the session cookie.

   a. The username, name, and avatar (picture) are still stored here even though @coverbeck suggested I remove them. I reasoning for keeping them was that their used in the view and still necessary (at least the avatar is, it's used to show the picture once you're logged in). I kept them all for consistency but we can take away if deemed necessary.

   b. The refresh token is now stored in the session cookie. It is encrypted. You gave a link for how to encrypt it, but I ended up taking a different path. I used AES encryption from the pycrypto library. This seems secure to me. My **only** security concern has to do with deployment. I'm not sure if by default the value from the previous deployment of is used again for the encryption key. My guess is that it's overwritten, in which case we're fine. But I haven't checked yet. Anyway if this *is* a vulnerability it would apply to any keys generated in deployment.

2. Support for a whitelist using bouncer is added!

Other stuff: 
- The weird and pointless get_logged_user code is removed
- unused imports are removed
- I had to require a more up to date version of urllib3 to avoid some weird errors from the default old version

jessebrennan committed:
Use session cookie instead of database

jessebrennan committed:
Get rid of more old db stuff

jessebrennan committed:
Get refresh token properly; Add some debugging logging

jessebrennan committed:
Set token attribute properly, add more debugging

jessebrennan committed:
Re order saving tokens

jessebrennan committed:
Make cookies secure

jessebrennan committed:
Encrypt refresh token; also add some debugging stuff

jessebrennan committed:
Store access_token, not all tokens

also remove some debug stuff

jessebrennan committed:
Add whitelist suport and some debug to comfirm

jessebrennan committed:
Update version of python2.7

Boto was crashing because a dependency required a slightly less
antiquated version of python. Boto is used by bouncer

jessebrennan committed:
Try adding a newer urllib3 dep

Hopefully this will resolve the weird boto error

jessebrennan committed:
Document why urllib3 is now a faux dependency

jessebrennan committed:
Remove outdated comment in dockerfile

jessebrennan committed:
Clean up fixme comments and print statements

coverbeck commented on app.py:
```diff
@@ -83,31 +77,109 @@ class ProdConfig(Config):
+        session as the one source of truth, but allow access and setting of
+        user properties here.
+        """
+        if user is not None:
```
I don't think you should put this information in the cookie. You should only need to put the access token and refresh token. This comment refers to the user, the email and the avatar, and to references further down.

See my response in reply to your PR comment.

coverbeck commented on app.py:
```diff
@@ -403,6 +470,23 @@ def login():
+        return user + '@' + '****'
+    return user + '@' + '***' + ('.' + tld)[-4:]
+
+
```
This is a good candidate for a test. :)

jessebrennan replied:
Agreed. Where would a test go? We aren't even running any right now, right?

coverbeck commented:
Regarding your PR overview/comments (good detail!):

1. Regarding email, avatar, etc., the problem is that you are not validating the access token so you will return that info even if the the user's tokens have expired. I'm torn as to whether you should leave it as an intermediate step, but I think the end goal should be to not have it.
2. Regarding the encryption:
   1.  My link was just the first thing I found when I Googled, and just a suggestion. It's fine to use a different library. 
    2. We would want the encryption key to persist between deploys, if by deploys you mean stopping and restarting. Otherwise, we wouldn't be able to decrypt cookies issued by the previous deploy.

coverbeck commented on app.py:
```diff
@@ -83,31 +77,109 @@ class ProdConfig(Config):
+        return cipher.decrypt(encrypted_token)
+
+    @refresh_token.setter
+    def refresh_token(self, value):
```
This is slick. 👍

jessebrennan replied:
Thanks!

coverbeck commented on app.py:
```diff
@@ -83,31 +77,109 @@ class ProdConfig(Config):
+    def refresh_token(self):
+        key = os.getenv('REFRESH_TOKEN_ENCRYPT_KEY')
+        if key is None:
+            raise ValueError('There is no encryption key set for the refresh token. '
```
Does this value error get surfaced to the client?

If so, I would rather that there not even be the hint of an encryption key to the client. Even though this is open source and a hacker could figure it out anyway, I would rather provide less info to the end user.

jessebrennan replied:
It's just a 500 error to the client. The exception is saved in the logs though.

coverbeck commented on app.py:
```diff
@@ -83,31 +77,109 @@ class ProdConfig(Config):
+
+    @property
+    def is_anonymous(self):
+        return self.email is None
```
If you get rid of email, which I think you should, then `self.access_token is None`.

coverbeck commented on app.py:
```diff
@@ -241,16 +305,16 @@ def check_session(cookie):
-                'email': logged_user.email,
-                'name': logged_user.name,
-                'avatar': logged_user.avatar
+                'error': 'No user is stored in the session. The user is not '
```
As we discussed, this is getting into #50, so not sure how far to go for this PR, but:

1. If current user is None and auth is not required, you should return `anonymous` as name.
2. If current user is None and auth is required, you should return a 401.
3. Also, this endpoint will go away, or maybe we can morph it into the `/me` endpoint.

I guess this is all optional at this point.

jessebrennan replied:
I'd rather put this off for the later PRs

coverbeck commented on app.py:
```diff
@@ -446,18 +530,18 @@ def callback():
+                b = Bouncer(whitelist)
+                if not b.is_authorized(email):
+                    return redirect(url_for('unauthorized', account=redact_email(email)))
+                else:
```
else without a body

coverbeck commented on app.py:
```diff
@@ -83,31 +77,109 @@ class ProdConfig(Config):
+        encrypted_token = session.get('refresh_token', None)
+        iv = session['iv']
+        cipher = AES.new(key, AES.MODE_CFB, iv)
+        return cipher.decrypt(encrypted_token)
```
Can you add some error handling and/or logging?

What if the encryption key gets changed and a cookie with the previous encrypted key comes in? The decryption will fail; will that get handled gracefully?

jessebrennan replied:
It won't get handled correctly, we just have to trust that we decrypted properly. Didn't you say we shouldn't be changing the key though?

What sort of behavior here do you think would be best? I think currently, google will just reject their bung token, right?

coverbeck replied:
We shouldn't be changing the key, but mistakes happen. :) Plus, if the key ever got compromised, we would intentionally change it.

How does decrypt behave if you give it the wrong key? Does it raise an exception? If so, I think you should return None.

coverbeck commented on app.py:
```diff
@@ -470,6 +554,7 @@ def callback():
 @app.route('/logout')
 @login_required
 def logout():
+    User().logout()
```
Can you not use current_user here? It has the `@login_required` annotation.

jessebrennan replied:
It would be the same. There's no reason not to except that it's already done this way. Do you think I should change it?

coverbeck replied:
If I understand correctly, doesn't the `@login_required` create a User instance, and then this code creates a second User instance?

I know the instances will be equivalent, but it seems suboptimal to create an unnecessary instance when you already have one you can use.

Unless I'm misunderstanding, I think you should change it.

coverbeck commented on app.py:
```diff
@@ -403,6 +470,23 @@ def login():
     return redirect(auth_url)
 
 
+def redact_email(email):
```
This is a good candidate for a unit test. :)

coverbeck reviewed (changes requested):
Overall, looking good! There's a lot of code and I have a lot of comments already, so I'm just going to Request Changes for now so you can get started. In further reviews, there may be comments regarding things I should have caught already, but hopefully not, or very few.

jessebrennan committed:
Use User() instead of current_user which defaults

to an anonymous user during the check_session call

jessebrennan committed:
Remove unused else

jessebrennan committed:
Examine the cookie passed to check_session

Before, I was expecting the session variable to be valid but it's not
because of the weird way this endpoint is hit. We need to check the
cookie in the state that it's passed in, and this means decoding it.

jessebrennan committed:
Test encryption and email redaction

coverbeck commented on requirements.txt:
```diff
@@ -11,3 +11,8 @@ oauth2client==4.1.2
 pyasn1==0.4.2
 psycopg2==2.7.4
 SQLAlchemy==1.1.5
+pycrypto>=2.6.1, <3
```
Please see if you can remove SQLAlchemy

coverbeck commented on app.py:
```diff
@@ -83,31 +78,99 @@ class ProdConfig(Config):
+
+    def logout(self):
+        """Clean up all the stuff we left in the session cookie"""
+        for attr in 'email', 'name', 'avatar', 'tokens', 'refresh_token':
```
Should be `access_token`, not `tokens`.

jessebrennan committed:
Add testing instructions to README

jessebrennan committed:
Fix clean-up bug and unused requirements

coverbeck commented on requirements.txt:
```diff
@@ -11,3 +11,8 @@ oauth2client==4.1.2
+cgp-bouncer>=0.1, <1
+# urllib3 is required b/c otherwise the old version causes a weird error with python 2.7.9
+# it is a requirement for botocore which is required by cgp-bouncer
+urllib3>=1.23, <2
```
Shouldn't you just pin these to specific versions, rather than ranges?

I know in the JS world, I have perfectly working builds start to fail because a newer version of a dependency got pulled in in a later build.

jessebrennan replied:
pinned!

jessebrennan committed:
Lock down dependencies

coverbeck commented on app.py:
```diff
@@ -446,18 +508,17 @@ def callback():
+            whitelist = os.getenv('EMAIL_WHITELIST_NAME')
+            # a value in the variable here is the flag for using a whitelist
+            if whitelist:
+                b = Bouncer(whitelist)
```
Would it be better to create just one Bouncer instance up front instead of creating a new instance on every login?

Just a thought, and I might be over-optimizing, especially if it's cheap to create the an instance.

coverbeck commented on requirements.txt:
```diff
@@ -3,11 +3,15 @@ elasticsearch-dsl==5.2.0
 pyasn1==0.4.2
 psycopg2==2.7.4
 SQLAlchemy==1.1.5
+pycrypto==2.6.1
```
You removed flask_sqlalchemy, but not SQLAlchemy, nor Flask-SQLAlchemy

coverbeck reviewed (changes requested):
Still a couple of SQL alchemy libraries in requirements.txt

jessebrennan committed:
Remove more old requirements

coverbeck reviewed (approved):
Please squash when merging. Good job!
